### ISSUE TITLE
Add default values

### ISSUE THREAD

**Issue by franky47** [47ng/next-usequerystate#197]
For when the query is not present, instead of returning `null`.

This should be done in a non-breaking way, as there are multiple use-cases for useQueryState:
- Where it makes sense to have an empty/nullish value (eg: filtering)
- Where it does not (eg: linkable controls)

Having a default value in React does not automatically update the URL to reflect it, which is debatable. If updating the URL, it would cause a side-effect depending on when the hook's parent component is mounted. If not, it breaks the idea that the URL is the source of truth (what if two hooks have different default values ?).

Use-cases in client code:
- https://github.com/rodmansw/rodmansw-web/blob/19891e049581addb3a3a3c5f1702196e4af373ca/pages/blog.js#L13
- https://github.com/maulanakurnia/personalwebsite/blob/1c719cb90184abc18d790a29ecd14a143188a94a/src/pages/blog.tsx#L31
- https://github.com/datacite/akita/blob/6c16461c89e8e72e276e72d4780f2c63bbc680c2/src/components/SearchPerson/SearchPerson.tsx#L56
- https://github.com/datacite/akita/blob/5a9f80e343951456129b47af3fc631dde4fa746c/src/pages/index.tsx#L9

**Comment by franky47** [comment]
It's possible to use function overloads to detect the use of a default value in the options, and update the return type accordingly.

Before:
<img width="364" alt="CleanShot 2021-03-01 at 14 02 37@2x" src="https://user-images.githubusercontent.com/1174092/109500731-e45e4b00-7a96-11eb-87fe-3f52b5a10e9b.png">

After:
<img width="365" alt="CleanShot 2021-03-01 at 14 02 59@2x" src="https://user-images.githubusercontent.com/1174092/109500739-e7f1d200-7a96-11eb-98d6-1a26eb79b398.png">

**Comment by iDavidB** [comment]
@franky47 Is there a reason as to why this hasn't been merged & released?

**Comment by franky47** [comment]
> @franky47 Is there a reason as to why this hasn't been merged & released?

Yes: I don't have time to work on this currently.

**Comment by iDavidB** [comment]
> Yes: I don't have time to work on this currently.

Ah, I thought your commit for this issue finished it.

**Comment by franky47** [comment]
I think there are some caveats that may need attention before providing such a solution, as mentioned in the comments up there, like how to deal with URL/state discrepancies when using default values.

If you have some feedback on that matter, I'd love to hear it!

**Comment by franky47** [comment]
Note: this was released in 1.4.0.

As for the behaviour with default values:
- If there is no query string present, the state will be set to the default value
- If there is a query string but it fails to parse, whatever the parser returns will be the state value (eg `NaN` for integer or float), the default value will not be returned there.
- When setting the state to `null`, the query key will be **removed** from the URL and the default value will be returned as the state value.
- The URL will never contain the default value (unless you manually set the state to it).

TL;DR: the default value only exists in the React tree as a fallback for when the query key is not present in the URL.
